iroh-rpc/src/client.rs

```diff
@@ -0,0 +1,331 @@
+    let (sender, receiver) = oneshot::channel();
+    command_sender
+        .send(Command::Dial {
+            peer_id,
+            peer_addr: addr,
+            sender,
+        })
+        .await
+        .expect("Receiver to not be dropped.");
+    match receiver.await.expect("Sender to not be dropped.") {
+        SenderType::Ack => Ok(()),
+        SenderType::Error(e) => Err(e),
+        _ => Err(RpcError::UnexpectedResponseType),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use libp2p::identity::Keypair;
+
+    use crate::config::RpcConfig;
+    use crate::handler;
+    use crate::rpc_from_config;
+    use crate::serde::{deserialize_request, serialize_response, Deserialize};
+    use crate::swarm;
+
+    #[derive(Deserialize, Serialize, Debug, Clone)]
+    struct GetRequest {
+        resource_id: String,
+    }
+
+    #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+    struct GetResponse {
+        response: String,
+    }
+
+    async fn get(
+        state: handler::State<String>,
+        _stream_id: Option<u64>,
+        param: Vec<u8>,
+    ) -> Result<Vec<u8>, RpcError> {
+        let _: GetRequest = deserialize_request(&param)?;
+
+        let bytes = serialize_response(GetResponse {
+            response: state.clone(),
+        })?;
+        Ok(bytes)
+    }
+
+    #[tokio::test]
+    async fn client_example() {
+        let a_state = handler::State::new(());
+        let a_keypair = Keypair::generate_ed25519();
+        let a_peer_id = a_keypair.public().to_peer_id();
+        let a_addr: Multiaddr = "/memory/1234".parse().unwrap();
+
+        let b_state = handler::State::new("Wooo!".to_string());
+        let b_addr: Multiaddr = "/memory/4321".parse().unwrap();
+        let b_keypair = Keypair::generate_ed25519();
+        let b_peer_id = b_keypair.public().to_peer_id();
+
+        let (mut a_client, a_server) = rpc_from_config(
+            RpcConfig::new()
+                .with_swarm(swarm::new_mem_swarm(a_keypair))
+                .with_state(a_state)
+                // can add the addrs here, or after the client has already
+                // been constructed
+                .with_addr("b", b_addr.clone(), b_peer_id),
+        )
+        .expect("rpc client to be created");
+        let (mut b_client, b_server) = rpc_from_config(
+            RpcConfig::new()
+                .with_swarm(swarm::new_mem_swarm(b_keypair))
+                .with_state(b_state)
+                .with_namespace("b", |n| n.with_method("get", get))
+                // can add the addrs here, or after the client has already
+                // been constructed
+                .with_addr("a", a_addr.clone(), a_peer_id),
+        )
+        .expect("rpc client to be created");
+
+        // run server
+        let a_thread_handle = tokio::spawn(async move {
+            a_server.run().await;
+        });
+        let b_thread_handle = tokio::spawn(async move {
+            b_server.run().await;
+        });
```

Comment:
Separated out the `Server` and `Client` again because I wasn't good enough at rust-foo to spawn this in its own thread without moving ownership of the `Client` (when the server was nested inside the client).

Seems to me we actually don't need access to the server, and could potentially create the server & set it running without needing to keep a reference to it anywhere (maybe just the JoinHandle of the thread running it), and potentially call `listen` and `dial` right when the client is created as well.

Also, this can be done in a future refactor once we've tried using it or in a higher level abstraction (eg `iroh-network-rpc`)

Thoughts?

Reply:
my sense is that the client only needs a channel to the server, but no other reference to it